model/component.js

```diff
@@ -0,0 +1,276 @@
+     * @param {string}  axis    The axis on which this span lies.
+     * @returns {boolean}
+     */
+    validateSpan(value, axis) {
+        switch (Component.testSpan(value)) {
+            case TestResult.DEFAULT:
+                console.log('Component: Field "' + axis + 'Span" is set to the default value.');
+                return false;
+
+            case TestResult.INVALID:
+                console.log('Component: Field "' + axis + 'Span" cannot be less than 1.');
+                return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Validates the entity field.
+     *
+     * Follows the rules of ParchKey's test string function.
+     *
+     * @since 1.0.0
+     *
+     * @see ParchKey.testStringValue
+     *
+     * @returns {boolean}
+     */
+    validateEntity() {
+        return ParchKey.testStringValue(this.entity, 'entity', 'Component') === TestResult.VALID;
+    }
+
+    /**
+     * Validate the ports field.
+     *
+     * Calls the Port validation function on each port in the array. The ports field cannot be null.
+     *
+     * @since 1.0.0
+     *
+     * @returns {boolean}
+     */
+    validatePorts() {
+        let valid = true;
+
+        if (!this.ports) {
+            console.log('Component: Field "ports" is set to the default value.');
+            return false;
+        }
+
+        this.ports.forEach(function(value, index, array) {
+            valid = value.validate() ? valid : false;
+
+            for (let i = index + 1; i < array.length; i++) {
+                // Check duplicate port labels
+                if (value.label === array[i].label) {
+                    valid = false;
+                    console.log('Component: Field "ports" has duplicate labels (' + value.label + ') at indices ' +
+                            index + ' and ' + i + '.');
+                }
+
+                // Check overlapping port locations
+                if (value.pos.is(array[i].pos)) {
+                    valid = false;
+                    console.log('Component: Field "ports" has overlapping ports at location ' + value.pos + ' at' +
+                            ' indices ' + index + ' and ' + i + '.');
+                }
+            }
+
+
+        });
+
+        // Check invalid port locations
+        for (let i = 0; i < this.ports.length; i++) {
+            // The coord validation checks for negative #s so we only need to check for greater than spans here.
+            if (this.ports[i].pos.x > this.xSpan || this.ports[i].pos.y > this.ySpan) {
+                valid = false;
+                console.log('Component: Field "ports" contains a port that exists outside of the component at ' +
+                        this.ports[i].pos + ' at index ' + i + '.');
+
+            } else if (!((this.ports[i].pos.x === 0 || this.ports[i].pos.x === this.xSpan) ||
+                    (this.ports[i].pos.y === 0 || this.ports[i].pos.y === this.ySpan))) {
```

Comment:
This case checks to make sure at least one of x or y are on the edge of the component. For example, only x needs to be on an edge and then y can be anywhere in between. I think if it is an && then we can only have ports at the corners.

Because this test only checks to make sure the port is on the edge, we need the test above to  make sure it doesn't go outside the bounding box of the component.